### PR TITLE
SONARJAVA-1745 Lookup supertypes of arguments for inference of method…

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 1]
```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
     List<JavaType> formalTypeSubstitutedTypes = formalType.typeSubstitution.substitutedTypes();
     TypeSubstitution result = substitution;
     if (argType.isParameterized()) {
-      List<JavaType> argTypeSubstitutedTypes = ((ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
-      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
-      result = mergeTypeSubstitutions(substitution, newSubstitution);
+      ParametrizedTypeJavaType parametrizedArgType = (ParametrizedTypeJavaType) argType;
+      if(parametrizedArgType.rawType == formalType.rawType) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [TypeInferenceSolver.java#L112](https://github.com/SonarSource/sonar-java/blob/2a6481ea19be947aa87d82767f05459da4bb90fc/java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java#L112): The Cyclomatic Complexity of this method "inferTypeSubstitutionInParameterizedType" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 2]
```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
...
+        List<JavaType> argTypeSubstitutedTypes = parametrizedArgType.typeSubstitution.substitutedTypes();
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+        return mergeTypeSubstitutions(substitution, newSubstitution);
+      }
+      JavaType superclass = argType.symbol.getSuperclass();
+      if(superclass != null) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: formatting: you should add a space between the "if" keyword and the parenthesis

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 3]
```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
...
+        superclass = typeSubstitutionSolver.applySubstitution(superclass, parametrizedArgType.typeSubstitution);
+        TypeSubstitution newSubstitution = inferTypeSubstitutionInParameterizedType(method, substitution, formalType, superclass, variableArity, remainingArgTypes);
+        if(!newSubstitution.substitutedTypes().isEmpty()) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: formatting: you should add a space between the "if" keyword and the parenthesis

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java (hunk 1 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: formatting: you should add a space between the "if" keyword and the parenthesis

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 4]
```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
...
+          result = mergeTypeSubstitutions(substitution, newSubstitution);
+        }
+      }
+      for (JavaType superInterface : argType.symbol.getInterfaces()) {
+        superclass = typeSubstitutionSolver.applySubstitution(superInterface, parametrizedArgType.typeSubstitution);
+        TypeSubstitution newSubstitution = inferTypeSubstitutionInParameterizedType(method, substitution, formalType, superclass, variableArity, remainingArgTypes);
+        if(!newSubstitution.substitutedTypes().isEmpty()) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: formatting: you should add a space between the "if" keyword and the parenthesis

File: java-frontend/src/test/files/sym/TypeInferenceOnSupertypes.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class A<T> {}
+class B<U> extends A<U> {}
+class Test {
+  B<String> b;
+  <S> S foo(A<S> a) {
+    foo(b); // should be resolved with infered substitution S -> String
+  }
+}
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/sym/TypeInferenceOnSupertypes.java]: missing new line.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 6]
```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
     List<JavaType> formalTypeSubstitutedTypes = formalType.typeSubstitution.substitutedTypes();
     TypeSubstitution result = substitution;
     if (argType.isParameterized()) {
-      List<JavaType> argTypeSubstitutedTypes = ((ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
-      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
-      result = mergeTypeSubstitutions(substitution, newSubstitution);
+      ParametrizedTypeJavaType parametrizedArgType = (ParametrizedTypeJavaType) argType;
+      if (parametrizedArgType.rawType == formalType.rawType) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [TypeInferenceSolver.java#L112](https://github.com/SonarSource/sonar-java/blob/58045b8aeb2c292ef868907786f0af78ba0913ef/java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java#L112): The Cyclomatic Complexity of this method "inferTypeSubstitutionInParameterizedType" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Pull request merged**